Show field values in category filter if it is linked to multiple fields

attekei opened the pull request:
Addresses #4909.

Summary of the problem:
- in dashboards, you expect to see in a category filter a list of field values if you have linked the filter to dashcard field(s) which are categories
- if all linked fields are the same, then you see the field values as expected
- if there are one or more different linked fields, you don’t see any field values (only a text input) which is unexpected

Some options for solving it (talked about these with @maz):
1) show field values of the first linked field (we have to define what “first linked field” means)
2) combine the field values of all linked fields and don’t eliminate duplicate values
3) combine the field values of all linked fields and eliminate duplicate values (i.e. of values overlap, don’t show a single value two times)

This PR experiments with approach (3) and it seems to work fairly well. I'm still curious to see if there are some any situations where sorting and removing duplicates of merged field values becomes a performance bottleneck.

This was also for me the first exercise of using `reselect` for a non-trivial use case. I'm interested to hear if there are ways to simplify the implementation.

Lacks tests but I want to discuss if the approach seems good before going further.

attekei committed:
First stab on merging field values from different fields

salsakran committed:
not sure Im a huge fan of this method signature, but whatevs. linters pass

salsakran commented on frontend/src/metabase/selectors/metadata.js:
```diff
@@ -123,8 +124,36 @@ export const getSegments = createSelector(
+
+// Check if the lengths of field value arrays equal for each field
+// TODO: Why can't we use plain shallowEqual, i.e. why the field value arrays change very often?
+const createFieldValuesEqualSelector = createSelectorCreator(defaultMemoize, (a, b) => {
```
what does "defaultMemoize" mean?

attekei replied:
It's the default memoization strategy of `reselect` which means that always when selector arguments change, the selector value is recalculated (as opposed to maintaining a cache of old selector argument <-> selector value combos and using cached values if possible, for instance).

Sadly the terminology of `reselect` isn't often very descriptive.

attekei replied:
I should probably add an explanatory comment to the code above, in minimum by adding a link to https://github.com/reactjs/reselect#customize-equalitycheck-for-defaultmemoize

salsakran commented on frontend/src/metabase/selectors/metadata.js:
```diff
@@ -123,8 +124,36 @@ export const getSegments = createSelector(
+// TODO Atte Keinänen 7/20/17: How should this work for remapped values?
+// TODO Atte Keinänen 7/20/17: Should we have any thresholds if the count of field values is high or we have many (>2?) fields?
+export const makeGetMergedParameterFieldValues = () => {
+    return createFieldValuesEqualSelector(getParameterFieldValuesByFieldId, (fieldValues) => {
```
why is this a function creator rather than just a function?

attekei replied:
If we just have a single function call of `createSelector` (or in this case the customized version called `createFieldValuesEqualSelector`) and export it, then we create only one selector that is reused everywhere in the app. 

This works well if all React components in the app call the selector with same parameters. In our case each instance of `<ParameterValueWidget>` expects different list of field values, and that list depends on `field_ids` array which `<ParameterValueWidget>` receives via React props.

We want that each `<ParameterValueWidget>` has its own selector because otherwise the parameters of one global selector change all the time and its never able to properly memoize anything. That leads to a terrible UI performance (I mean, it really is terrible, I tested it).

I should probably add an explanatory comment to the code above, in minimum by adding a link to https://github.com/reactjs/reselect#sharing-selectors-with-props-across-multiple-components and explaining that `fields_ids` value is specific to each ParameterValueWidget instance.

attekei replied:
And some clarifications:

- The old approach was fast because it basically just used the field values list that the backend returned as-is. In that case any `reselect` stuff was useless.

- The new approach does a relatively heavyweight merge by sorting and then removing duplicates of combined field value lists (one list per each field id). That's why without `reselect` and the equality comparison / selector creator nastiness the UI performance becomes terrible.

attekei committed:
Merge branch 'release-0.26.0' into fix-4909, add commentary

attekei committed:
More commentary about field value selectors

attekei committed:
Fix sorting with remapped values

attekei commented:
This should behave now correctly for remapped values as well (in terms of the sorted order of field values etc).

I did some performance auditing, and with 5 different fields combined (which is a pretty crazy scenario) the selector calculation took around 15ms. With 2 different fields with maximum number of field values in both it was around 5ms. So shouldn't be too scary.

I will still write unit tests for `makeGetMergedParameterFieldValues`, then I think that this in a pretty good state! I btw added more commentary as well.

attekei committed:
Add tests, fix issue with public/embedded dashboard filters

attekei commented:
Added tests and fixed an issue with public dashboards (that one test case for them caught!). Do you think @salsakran that this is in a mergeable state?

attekei committed:
Minor cleanup for eslint

attekei committed:
Fix unit tests

salsakran commented on frontend/src/metabase/selectors/metadata.js:
```diff
@@ -123,22 +125,69 @@ export const getSegments = createSelector(
+// TODO Atte Keinänen 7/20/17: Should we have any thresholds if the count of field values is high or we have many (>2?) fields?
+export const makeGetMergedParameterFieldValues = () => {
+    return createFieldValuesEqualSelector(getParameterFieldValuesByFieldId, (fieldValues) => {
+        const fieldIds = Object.keys(fieldValues)
```
I think using "fieldIds" here is confusing. For the longest time I thought these were literal field IDs (eg, `123`) rather than field values that have not been remapped.

attekei replied:
they are fieldIds – your confusion probably originates from terminology ambiguity between "parameter mapping" and "remapping" which are two different things ("parameter mapping" refers to how dashcards map to filters whereas "remapping" means replacing shown field values with something new)

Will try to clarify that in source code

salsakran commented on frontend/src/metabase/selectors/metadata.js:
```diff
@@ -123,22 +125,69 @@ export const getSegments = createSelector(
+        if (fieldIds.length === 0) {
+            // If we have no mapped fields, then don't return any values
+            return [];
+        } else if (fieldIds.length === 1) {
```
this seems fragile. Couldn't there be other fields that have a single valid field value that aren't booleans?

attekei replied:
again, it means that there is just a single field, not single field value

salsakran commented on frontend/src/metabase/selectors/metadata.js:
```diff
@@ -123,22 +125,69 @@ export const getSegments = createSelector(
+            const sortedMergedValues = _.chain(Object.values(fieldValues))
+                .flatten(true)
+                // Use remapped value for sorting if it is available
+                .sortBy(fieldValue => fieldValue.length === 2 ? fieldValue[1] : fieldValue[0])
```
This smells off to me. I'd prefer a variable "isRemapped" that encapsulates this logic without needing a comment.

attekei replied:
yup, I extracted that to a variable

salsakran reviewed (changes requested):
Some minor naming changes, and this will be good!

attekei committed:
Fix issue with sql template tag filters

attekei committed:
Add commentary, extract inlined expressions to variables

attekei commented:
I did the changes requested by @salsakran so I'll merge this now!
